library/src/main/scala/scala/scalajs/concurrent/QueueExecutionContext.scala

```diff
@@ -1,22 +1,55 @@
 package scala.scalajs.concurrent
 
 import scala.concurrent.ExecutionContextExecutor
-import scalajs.js
 
-private[concurrent] object QueueExecutionContext
-    extends ExecutionContextExecutor {
+import scala.scalajs.js
+import scala.scalajs.js.|
 
-  def execute(runnable: Runnable): Unit = {
-    js.timers.setTimeout(0) {
-      try {
-        runnable.run()
-      } catch {
-        case t: Throwable => reportFailure(t)
-      }
+object QueueExecutionContext {
+  def timeouts(): ExecutionContextExecutor =
+    new TimeoutsExecutionContext
+
+  def promises(): ExecutionContextExecutor =
+    new PromisesExecutionContext
+
+  def apply(): ExecutionContextExecutor =
+    if (js.isUndefined(js.Dynamic.global.Promise)) timeouts()
+    else promises()
+
+  private final class TimeoutsExecutionContext extends ExecutionContextExecutor {
+    def execute(runnable: Runnable): Unit = {
+      js.Dynamic.global.setTimeout({ () =>
```

Comment:
Also, `setTimeout` is not *supposed* to be be global state. Other than mocking, no reasonable code should ever change `setTimeout`.

By the same argument, we should capture `Math` in `java.lang.Math`, `Date` in `java.util.Date`, `Object` in ... er, boom! So many places we should capture `Object` ...